stdlib/public/Concurrency/TaskGroup.cpp

```diff
@@ -0,0 +1,303 @@
+//===--- TaskGroup.cpp - Task Group internal message channel ------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Object management for async child tasks that are children of a task group.
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Runtime/Concurrency.h"
+#include "swift/ABI/Task.h"
+#include "swift/ABI/Metadata.h"
+#include "swift/Runtime/HeapObject.h"
+#include "TaskPrivate.h"
+
+using namespace swift;
+using GroupFragment = AsyncTask::GroupFragment;
+using FutureFragment = AsyncTask::FutureFragment;
+
+using ReadyQueueItem = GroupFragment::ReadyQueueItem;
+using ReadyStatus = GroupFragment::ReadyStatus;
+using GroupPollResult = GroupFragment::GroupPollResult;
+
+// =============================================================================
+// ==== destroy ----------------------------------------------------------------
+
+// FIXME: implement this properly, unsure if this is correct
+void GroupFragment::destroy() {
+  // TODO: need to release all waiters as well
+//  auto waitHead = waitQueue.load(std::memory_order_acquire);
+//  switch (waitHead.getStatus()) {
+//    case GroupFragment::WaitStatus::Waiting:
+//      assert(false && "destroying a task group that still has waiting tasks");
+//  }
+
+  mutex.lock(); // TODO: remove fragment lock, and use status for synchronization
+  // Release all ready tasks which are kept retained, the group destroyed,
+  // so no other task will ever await on them anymore;
+  ReadyQueueItem item;
+  bool taskDequeued = readyQueue.dequeue(item);
+  while (taskDequeued) {
+    swift_release(item.getTask());
+    bool taskDequeued = readyQueue.dequeue(item);
+  }
+  mutex.unlock(); // TODO: remove fragment lock, and use status for synchronization
+}
```

Comment:
Unsure if anything else to cleanup here; the wait queue never has values in it, just waiting tasks; so nothing to cleanup there I think